samza-core/src/main/java/org/apache/samza/container/grouper/task/GroupByContainerCount.java

```diff
@@ -74,7 +74,7 @@ public GroupByContainerCount(int containerCount) {
     // Convert to a Set of ContainerModel
     Set<ContainerModel> containerModels = new HashSet<>();
     for (int i = 0; i < containerCount; i++) {
-      containerModels.add(new ContainerModel(i, taskGroups[i]));
+      containerModels.add(new ContainerModel(String.valueOf(i), taskGroups[i]));
```

Comment:
This line, and I believe a few others, assume the container/processor ids are still a sequence of integers (even though the type is String). 

I see how this patch ensures that this code is only invoked in the legacy case, where that would be true. But this precondition/assumption is now subtle; it's not obvious by looking at the types. 

I suggest that we at least add a validation with a clear exception that ensures that the external strings (e.g. those that come from the TaskAssignmentManager) are all parseable integers.